File: Topics/Binary/Search2DMatrix.java
package Topics.Binary;

public class Search2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }
        int m = matrix.length, n = matrix[0].length;
        int l = 0, r = m*n-1;
        while (l <= r) {
            int mid = l + (r-l)/2;
            if (matrix[mid/n][mid%n] == target) {
                return true;
            }
            if (matrix[mid/n][mid%n] > target) {
                r = mid-1;
            } else {
                l = mid+1;
            }
        }
        return false;
    }

    /* Integers in each row are sorted in ascending from left to right. */
    /* Integers in each column are sorted in ascending from top to bottom. */
    public boolean searchMatrixII(int[][] matrix, int target) {
        if (matrix.length == 0) {
            return false;
        }
        int r = 0, c = matrix[0].length-1;
        while (r < matrix.length && c >= 0) {
            if (matrix[r][c] == target) {
                return true;
            }
            if (matrix[r][c] > target) {
                c--;
            } else {
                r++;
            }
        }
        return false;
    }
}

File: Topics/Design/al/Search_2DMatrix.cpp
class Solution {
public:
    // 每行有序，且下一行第一个元素 > 上一行最后一个元素
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty() || matrix[0].empty()) {
            return false;
        }

        int m = matrix.size();
        int n = matrix[0].size();

        int left = 0, right = m * n - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            int value = matrix[mid / n][mid % n];

            if (value == target) {
                return true;
            } else if (value < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return false;
    }

    // 每行递增，每列递增
    bool searchMatrixII(vector<vector<int>>& matrix, int target) {
        if (matrix.empty() || matrix[0].empty()) {
            return false;
        }

        int rows = matrix.size();
        int cols = matrix[0].size();

        int r = 0;
        int c = cols - 1;

        while (r < rows && c >= 0) {
            if (matrix[r][c] == target) {
                return true;
            } else if (matrix[r][c] > target) {
                --c;      // 去掉一整列
            } else {
                ++r;      // 去掉一整行
            }
        }

        return false;
    }
};
